src/vm_pages.rs

```diff
@@ -826,6 +898,52 @@ impl<T: GuestStagePageTable> VmPages<T, VmStateFinalized> {
     ) -> ActiveVmPages<T> {
         ActiveVmPages::new(self, vmid, prev_tlb_version)
     }
+
+    /// Maps num_pages of non-confidential 4Kb pages starting at page_addr. The range must fit in
+    /// the pre-desiginated range set by `set_shared_page_range` and there must be no overlap between
+    /// the ranges for confidential and non-confidential pages
+    pub fn insert_shared_pages(
+        &self,
+        from_addr: GuestPageAddr,
+        num_pages: u64,
+        from: &VmPages<T, VmStateFinalized>,
+        guest_addr: GuestPageAddr,
+    ) -> Result<()> {
+        let map_range =
+            PageRangeInfo::new(guest_addr.bits(), num_pages).ok_or(Error::AddressOverflow)?;
+        let (shared_list, mapper) = {
+            let shared_page_range = self.shared_page_range.lock();
+            if let Some(shared_page_range) = shared_page_range.deref() {
+                if map_range.base < shared_page_range.base
+                    || map_range.limit > shared_page_range.limit
+                {
+                    return Err(Error::SharedPageNotInRange);
+                }
+            } else {
+                return Err(Error::SharedPageRangeNotSet);
+            }
+            let confidential_page_range = self.confidential_page_range.lock();
+            if let Some(confidential_page_range) = confidential_page_range.deref() {
+                if confidential_page_range.does_range_overlap(&map_range) {
+                    return Err(Error::SharedPageRangeOverlap);
+                }
+            }
+
+            let shared_list = from
+                .root
+                .get_shareable_range::<Page<Shareable>>(from_addr, PageSize::Size4k, num_pages)
+                .map_err(|_| Error::SharedPageNotMapped)?;
+
+            let mapper = VmPagesMapper::new_shareable(self, guest_addr, num_pages)?;
+            (shared_list, mapper)
+        };
+
+        for (page, addr) in shared_list.zip(guest_addr.iter_from()) {
```

Comment:
The state transition from `Mapped` to `Shared` (or bumping of the `Shared` refcount) probably shouldn't happen until we've done all the setup work above. Something like:

```
for (page, addr) in shared_list.zip(guest_addr.iter_from()) {
    let shared = page_tracker.share_page(page).unwrap();
    mapper.map_page(addr, shared).unwrap();
}
```

Reply:
Not resolved; see my other comment about splitting this into two steps so that we don't bump the refcount until we've guaranteed that the page will actually get mapped.

Reply:
Believe this is resolved per our conversation.